proposals/0568-specrec.md

```diff
@@ -0,0 +1,247 @@
+---
+author: Andreas Klebinger
+date-accepted: ""
+ticket-url: ""
+implemented: ""
+---
+
+This proposal is [discussed at this pull request](https://github.com/ghc-proposals/ghc-proposals/pull/568).
+
+# Transitive specialization support.
+
+Implement a `{-# SPECREC foo #-}` pragma. Using such a pragma on a function will encourage GHC to specialize
+all uses of `foo`, expose foos unfolding **and also do so for all use sites of `foo`** in a recursive fashion.
+
+That is if we have
+
+```haskell
+{-# SPECREC foo #-}
+elem :: Eq a => a -> [a] -> Bool
+elem = ...
+
+elems :: (Foldable t, Eq a) => a -> t [a] -> Bool
+elems x = any (elem x)
+
+foo :: [String]
+foo xs = elems 'c'
+```
+
+It will act as if we had written:
+
+```haskell
+{-# SPECREC foo #-}
+elem :: Eq a => a -> [a] -> Bool
+elem = ...
+
+{-# SPECREC foo #-}
+elems :: (Foldable t, Eq a) => a -> t [a] -> Bool
+elems x = any (elem x)
```

Comment:
In the “as if” section I’d expect `SPECIALIZE`, not the `SPECREC` you are introducing.

Reply:
I've expanded on the differences.